openff/recharge/esp/qcarchive.py

```diff
@@ -0,0 +1,402 @@
+        for shell in wavefunction.basis.center_data[atom].electron_shells
+        for angular_momentum in shell.angular_momentum
+    }
+
+    spherical_maps = {
+        L: numpy.array(
+            list(range(L * 2 - 1, 0, -2)) + [0] + list(range(2, L * 2 + 1, 2))
+        )
+        for L in angular_momenta
+    }
+
+    # Build a flat index that we can transform the AO quantities
+    ao_map = []
+    counter = 0
+
+    for atom in wavefunction.basis.atom_map:
+
+        center = wavefunction.basis.center_data[atom]
+        for shell in center.electron_shells:
+
+            if shell.harmonic_type == "cartesian":
+                ao_map.append(numpy.arange(counter, counter + shell.nfunctions()))
+
+            else:
+                smap = spherical_maps[shell.angular_momentum[0]]
+                ao_map.append(smap + counter)
+
+            counter += shell.nfunctions()
+
+    ao_map = numpy.hstack(ao_map)
+
+    reverse_ao_map = {map_index: i for i, map_index in enumerate(ao_map)}
+    reverse_ao_map = numpy.array([reverse_ao_map[i] for i in range(len(ao_map))])
+
+    reordered_density = density[reverse_ao_map[:, None], reverse_ao_map]
+    return reordered_density
+
+
+@requires_package("psi4")
+def _compute_esp(
+    qc_molecule, density, esp_settings, grid
+) -> Tuple[numpy.ndarray, numpy.ndarray]:
+    """Computes the ESP and electric field for a particular molecule on
+    a specified grid and using the specified settings.
+
+    Parameters
+    ----------
+    qc_molecule
+        The molecule to compute the ESP / electric field of.
+    density
+        The electron density of the molecule.
+    esp_settings
+        The settings to use when computing the ESP / electric field.
+    grid
+        The grid to evaluate the ESP and electric field on.
+
+    Returns
+    -------
+        A tuple of the evaluated ESP with shape=(n_grid_points, 1) and the electric
+        field with shape=(n_grid_points, 3)
+    """
+    import psi4
+
+    psi4_molecule = psi4.geometry(qc_molecule.to_string("psi4", "angstrom"))
+    psi4_molecule.reset_point_group("c1")
+
+    psi4_wavefunction = psi4.core.RHF(
+        psi4.core.Wavefunction.build(psi4_molecule, esp_settings.basis),
+        psi4.core.SuperFunctional(),
+    )
+    psi4_wavefunction.Da().copy(psi4.core.Matrix.from_array(density))
+
+    psi4_calculator = psi4.core.ESPPropCalc(psi4_wavefunction)
+    psi4_grid = psi4.core.Matrix.from_array(grid)
+
+    esp = numpy.array(
+        psi4_calculator.compute_esp_over_grid_in_memory(psi4_grid)
+    ).reshape(-1, 1)
+
+    # TODO: enable this once an updated psi4 is released.
+    # field = numpy.array(psi4_calculator.compute_field_over_grid_in_memory(psi4_grid))
+
+    field = numpy.zeros((len(esp), 3))
+
+    return esp, field
+
+
+@requires_package("qcportal")
+@requires_package("cmiles")
+def from_qcarchive(
+    data_set_name: str,
+    subset: Optional[Iterable[str]],
+    esp_settings: ESPSettings,
```

Comment:
Should there be some check that the basis/method in ESPSettings matches the basis/method using for the QCArchive calculation or are we trusting that the user has checked they are the same?

Reply:
I think this is being handled when we query the results from fractal - we explicitly filter based on method + basis and then check to make sure we still have results returned for all molecule ids!

Reply:
Ah yes I missed that you pass it into the get results call good job.